syntax = "proto3";
package bombplane_proto; //定义protobuf的包名称空间，对应C++,C#的namespace，Java的package

// request		-- need reply(response)
// notification	-- no reply needed
// broadcast	-- no reply needed

enum TYPE
{
	Unknown						= 0;


    // Login & Reg

 	Login_Request  				= 10001;
 	Login_Response  			= 10002;

	Onlinelist_Notification		= 10003; // S->C: 为新登录的用户发送列表
	Updateonline_Broadcast		= 10004; // 上线或下线需要对所有用户发送

	// Reg_Request 				= 10011;
	// Reg_Response				= 10012;


	// Invite

	Invite_Request 				= 10021; // A点击B进行邀请，服务器只转发
	Invite_Response				= 10022; // B接收A的邀请
	Updateroom_Broadcast		= 10023; // 服务器对所有用户发送“AB进入或退出房间”的消息


	// KeepAlive

	Quit_Notification			= 20001; // C->S: 关闭窗口或掉线

	Keepalive_Request			= 20002;
	Keepalive_Response 			= 20003;


	// Game

	Initpos_Request				= 30001; // 初始化设置飞机坐标
	Gamestart_Response			= 30002; // 通知双方游戏开始

	Gameover_Notification		= 30003; // 输赢结束
	Gamecrush_Notification		= 30004; // 异常结束

	Bomb_Request				= 30011; // 单坐标轰炸
	Bomb_Response				= 30012; // 向双方同时发送轰炸结果

	Guess_Request				= 30021; // 三坐标（整机）猜测
	Guess_Response				= 30022; // 向双方同时发送猜测结果
}

/*下面定义具体的消息内容，Type枚举中的每个消息ID，如果有消息体，则会对应一个message 定义，如果无消息体则不必要*/ 

// Login & Reg

message LoginRequest
{
	bytes username	= 1;
	bytes password	= 2;
};



enum LOGINSTATE
{
	SUCCESS			= 0;
	SUCCESS_KICK	= 1;

	NOT_EXIST		= 3;
	WRONG_PASSWORD	= 4;
	SERVER_ERROR	= 5;
};

message LoginResponse
{
	LOGINSTATE state	= 1;
	int32 userid		= 2;
};



message OnlineUser
{
	bytes username	= 1;
	int32 userid	= 2;
	bool  available	= 3;
};

message OnlinelistNotification
{
 	repeated OnlineUser onlinelist	= 1;
};


message UpdateonlineBroadcast
{
 	bytes username	= 1;
 	int32 userid	= 2;
 	bool  onoff		= 3;
};


// Invite

message InviteRequest
{
	int32 taruserid	= 1;
	int32 srcuserid	= 2;
};

message InviteResponse
{
	int32 taruserid	= 1;
	bool  accept	= 2;
};


message UpdateroomBroadcast
{
	int32 userid1	= 1;
	int32 userid2	= 2;
	bool  inout		= 3;
};



// Game

message Coordinate
{
	int32 x = 1;
	int32 y = 2;
};

message PlaneLocator
{
	Coordinate pos1 = 1;
	Coordinate pos2 = 2;
	Coordinate pos3 = 3;
};

message InitposRequest
{
	PlaneLocator loc	= 1;
	int32 userid		= 2;
};

message GameoverNotification
{
	int32 winnerid = 1;
};


// enum CRUSH_REASON
// {
// 	PEER_DISCONNECT = 0;
// 	SERVER_ERROR 	= 1;
// }
// message GamecrushNotification
// {
// 	CRUSH_REASON reason = 1;
// }

message BombRequest
{
	Coordinate pos	= 1;
	int32 userid	= 2;
};

enum BOMB_RESULT
{
	MISS 		= 0;
	HIT 		= 1;
	DESTORYED 	= 2;
};

message BombResponse
{
	BOMB_RESULT res	= 1;
	Coordinate pos	= 2;
};


message GuessRequest
{
	PlaneLocator loc	= 1;
	int32 userid		= 2;
};


message GuessResponse
{
	bool res			= 1;
	PlaneLocator loc	= 2;
}


/*请求消息集合，任何时刻根据MSG 的id值，最多只有一个有效性， 从程序的逻辑上去保证，编译器（不管是protoc还是具体语言的编译器都无法保证）*/

message Body
{
    // Login & Reg

 	LoginRequest loginrequest						= 1;
 	LoginResponse loginresponse						= 2;

	OnlinelistNotification onlinelistnotification	= 3;
	UpdateonlineBroadcast updatelistnotification	= 4;


	InviteRequest inviterequest						= 5;
	UpdateroomBroadcast updateroombroadcast			= 6;

	// Game

	InitposRequest initposrequest					= 7;

	GameoverNotification gameovernnotification		= 8;
	// GamecrushNotification gamecrushnotification	= 14;

	BombRequest bombrequest							= 9;
	BombResponse bombresponse						= 10;

	GuessRequest guessrequest						= 11;
	GuessResponse guessresponse						= 12;
}


message Message
{
	TYPE type = 1;
	// fixed32 sequence = 2; 
	// 消息系列号，主要用于Request和Response，Response的值必须和Request相同，使得发送端可以进行事务匹配处理

	Body body = 2;
}

